### ISSUE TITLE
Sometimes on Mac the menu bar icon doesn't show up on the menu bar of the display without the focus until it get the focus

### ISSUE THREAD

**Issue by Draykan** [simonrob/email-oauth2-proxy#174]
Sometimes on macOS 13 Ventura, when using an external display as an extended desktop and launching Email OAuth 2.0 Proxy, the menu bar icon is correctly added only on the display (external or integrated) with the focus when the proxy is started.
An empty space to accomodate the proxy icon is added between the other icons on the menu bar of the display without the focus but the icon is actually displayed only when it gets the focus.
This is true both when starting the proxy manually with `python3 emailproxy.py` and when restarted with `launchctl start ac.robinson.email-oauth2-proxy` when configured to run automatically as a Launch Agent.
As it doesn't always happen, it seems that the best way to trigger the issue is to launch the proxy from a Terminal window on a fresh new Desktop space and look at the menu bar of the current Desktop on the other display.
If needed, I can register a video to show the glitch.

System information and software versions:
• MacBook Air (M1, 2020)
• LG UltraFine Display 21,5" 22MD4KA 
• macOS Ventura 13.5 (22G5048d) Beta 3
• Command Line Tools for Xcode 14.3 (installed for the first time on 04/04/2023 and according to System Information app updated to the same version on 06/06/2023 with Software Update)

`python3 --version`
```
Python 3.9.6
```

`python3 -m pip list`
```
Package                              Version
------------------------------------ -------
altgraph                             0.17.2
bottle                               0.12.23
cffi                                 1.15.1
cryptography                         39.0.0
future                               0.18.2
macholib                             1.15.2
Pillow                               9.4.0
pip                                  21.2.4
prompt-toolkit                       3.0.36
proxy-tools                          0.1.0
pycparser                            2.21
pyobjc-core                          9.2
pyobjc-framework-Cocoa               9.2
pyobjc-framework-Quartz              9.0.1
pyobjc-framework-Security            9.2
pyobjc-framework-SystemConfiguration 9.0.1
pyobjc-framework-WebKit              9.0.1
pyoslog                              1.1.0
pystray                              0.19.4
pywebview                            4.1
setuptools                           58.0.4
six                                  1.15.0
timeago                              1.0.16
wcwidth                              0.2.6
wheel                                0.37.0
WARNING: You are using pip version 21.2.4; however, version 23.1.2 is available.
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
```

**Comment by simonrob** [comment]
Thanks for reporting this – yes, this is something I've noticed occasionally with [pystray](https://github.com/moses-palmer/pystray) (which provides the proxy's menu bar icon functionality), but I haven't been able to identify the exact conditions that make it occur. Restarting the proxy fixes this in all instances I've encountered.

If you're able to isolate the cause in a reproducible manner then I'm happy to investigate further.

**Comment by Draykan** [comment]
It seems I can always reproduce it at will with these steps:

1. Connect an external display to the MacBook.
2. Set the external display as an extended desktop and the internal display as main/principal.
3. If running, close Email OAuth 2.0 Proxy.
4. Focus on the internal display, activate Mission Control (three fingers up gesture) and add a new empty Desktop on the external display by clicking the + symbol.
5. Click on the new Desktop thumbnail on the external display to switch to it. Please note that the focus will NOT switch to the external display.
6. Activate Mission Control (three fingers up gesture) and add a new empty Desktop on the internal display by clicking the + symbol.
7. Click on the new Desktop thumbnail on the internal display to switch to it.
8. Open a Terminal window on the new empty Desktop od the internal display.
9. Launch Email OAuth 2.0 Proxy with `python3 emailproxy.py` or `launchctl start ac.robinson.email-oauth2-proxy` if it is set to run automatically as a Launch Agent.

This is a screenshot of the issue, the icon should be between the cloud and shield icons.
<img width="757" alt="Screenshot of the missing icon" src="https://github.com/simonrob/email-oauth2-proxy/assets/151580/69de6738-3d0f-4035-b891-b7eafd2abdd1">

During  my test I've also been able for a couple of times to get an icon like this! :))
<img width="755" alt="Screenshot of the truncated icon" src="https://github.com/simonrob/email-oauth2-proxy/assets/151580/ccc0a267-9281-49d1-a78c-63d3f3aaf9d4">

**Comment by simonrob** [comment]
Thanks for the detailed follow-up. This doesn't trigger the issue for me, unfortunately.

However, given you seem to have a reproducible setup, it is worth trying to confirm whether this is solely an issue with pystray, or whether the proxy's minor modifications to it (high-resolution icon, refreshable menus, inline icons) have anything to do with this.

Could you try your set of steps with a [basic pystray example](https://pystray.readthedocs.io/en/latest/usage.html#creating-a-system-tray-icon) and see whether it still occurs?

**Comment by Draykan** [comment]
Hi Simon,
I used the example at the link you provided and it seems that I cannot reproduce the issue with basic pystray:
<img width="832" alt="Screenshot of the test icon" src="https://github.com/simonrob/email-oauth2-proxy/assets/151580/fe2ec05e-9cd7-4adc-9c22-e47286914ec7">

Feel free to ask me to perform on my machine any test you like.

**Comment by simonrob** [comment]
Thanks for the offer – I will dig a little into this and see whether I can find out the root cause here.

Out of interest, when the icon is missing, is it actually there but just not visible? I.e., if you click on the space where it should be, do you see a slightly blurry version of what you'd normally expect?

**Comment by Draykan** [comment]
I think I found something interesting.
In the last few days I updated some python packages on my Mac and today I noticed that I couldn't reproduce the problem anymore. So I downgraded the updated packages one by one to the version I had when I posted this issue and I found that with [Pillow](https://github.com/python-pillow/Pillow) 9.4.0 the problem is there but with version 9.5.0 I cannot reproduce it.
Maybe you tried to reproduce the issue with Pillow 9.5.0?

• Python packages versions triggering the issue:
`python3 -m pip list`
```
Package                              Version
------------------------------------ -------
altgraph                             0.17.2
bottle                               0.12.25
cffi                                 1.15.1
cryptography                         41.0.1
future                               0.18.2
macholib                             1.15.2
Pillow                               9.4.0 <--
pip                                  21.2.4
prompt-toolkit                       3.0.38
proxy-tools                          0.1.0
pycparser                            2.21
pyobjc-core                          9.2
pyobjc-framework-Cocoa               9.2
pyobjc-framework-Quartz              9.2
pyobjc-framework-Security            9.2
pyobjc-framework-SystemConfiguration 9.2
pyobjc-framework-WebKit              9.2
pyoslog                              1.1.0
pystray                              0.19.4
pywebview                            4.2.2
setuptools                           58.0.4
six                                  1.15.0
timeago                              1.0.16
typing_extensions                    4.6.3
wcwidth                              0.2.6
wheel                                0.37.0
```

• Python package versions not triggering the issue:
`python3 -m pip list`
```
Package                              Version
------------------------------------ -------
altgraph                             0.17.2
bottle                               0.12.25
cffi                                 1.15.1
cryptography                         41.0.1
future                               0.18.2
macholib                             1.15.2
Pillow                               9.5.0 <--
pip                                  21.2.4
prompt-toolkit                       3.0.38
proxy-tools                          0.1.0
pycparser                            2.21
pyobjc-core                          9.2
pyobjc-framework-Cocoa               9.2
pyobjc-framework-Quartz              9.2
pyobjc-framework-Security            9.2
pyobjc-framework-SystemConfiguration 9.2
pyobjc-framework-WebKit              9.2
pyoslog                              1.1.0
pystray                              0.19.4
pywebview                            4.2.2
setuptools                           58.0.4
six                                  1.15.0
timeago                              1.0.16
typing_extensions                    4.6.3
wcwidth                              0.2.6
wheel                                0.37.0
```

To answer your question, when I click on the space where the icon should be, that (internal or external) display gets the focus and the menu icon appears.
Please see this screen recording:
https://github.com/simonrob/email-oauth2-proxy/assets/151580/7a7c2bd5-a7a5-4b04-bb9d-465f678231b3

**Comment by simonrob** [comment]
Thanks for following up – that's a useful finding. I still can't replicate this issue reliably, though, regardless of the Pillow version. I suspect there's something else going on because you weren't able to reproduce the effect with the basic pystray example that also uses Pillow to create an icon (assuming this was before the version upgrade).

On an older version of macOS I can [force a similar effect](https://github.com/simonrob/email-oauth2-proxy/assets/934006/c4b1e08e-0827-44e4-b8c5-aa0957cd2476) by removing the proxy's modifications to the pystray icon, but can't get anything like this to happen with the unmodified proxy. Realistically, I think that given that the issue fixes itself, and also seems to have been resolved entirely by an updated dependency, it's best to just accept this quirk for now. But thanks again for your efforts in trying to diagnose the cause.

**Comment by Draykan** [comment]
I've just checked again with the basic pystray example and Pillow 9.4.0 and this time I've been able to reproduce the issue.
Probably when I tested it I gave up too quickly because I was sure I was still using Pillow 9.4.0 at the time.
[This](https://github.com/simonrob/email-oauth2-proxy/assets/151580/77abbee3-39c3-4593-bc34-69ddd7a4dc86) is the screen recording of Email Proxy and the pystray example simultaneously having the issue.
I still confirm that with Pillow 9.5.0 I cannot reproduce the issue anymore.

**Comment by simonrob** [comment]
Glad to hear this – it means that the proxy is unlikely to be the culprit, and there was some issue resolved in Pillow 9.5.0.

The [version comparison](https://github.com/python-pillow/Pillow/compare/9.4.0...9.5.0) and [changelog](https://pillow.readthedocs.io/en/stable/releasenotes/9.5.0.html) don't immediately reveal anything that might be the culprit, so I think it's unlikely that this can be worked around in the proxy. Probably, unless you have any reason not to, simply updating Pillow is the best resolution.

**Comment by simonrob** [comment]
I'll close this now given there's probably nothing that can be done. Feel free to reopen if you discover anything new, or if the problem returns with Pillow 9.5.0+.
